Advanced adventure for Watson + Swift room

jkingoliver opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gameontext/gameon-gitbook/65)
<!-- Reviewable:end -->

BarDweller commented:
Reviewed 8 of 11 files at r2.
Review status: 8 of 11 files reviewed at latest revision, 18 unresolved discussions.

---

*[walkthroughs/addWatsonConversation.adoc, line 10 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV03U5l7mjmXtPBVfv:-KgV03U5l7mjmXtPBVfw:b-7ogkwp) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L10)):*
> ```AsciiDoc
> :imagesdir: /images
> 
> Where we learn about how to integrate Watson Conversation service with a Swift chat room.
> ```

learn about how to integrate the Watson Conversation service
or
learn how to integrate Watson Conversation service

Something doesn't sit right with it worded as is.. I think it's the 'about' that seems superfluous .. 

---

*[walkthroughs/addWatsonConversation.adoc, line 22 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV0cL0Y1rnX5Vv9V6o:-KgV0cL1o0LZ_WzDCaXK:b-ey86zq) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L22)):*
> ```AsciiDoc
> More and more developers are demanding the use of cognitive features in their applications. That’s why IBM Watson is targeting the need by providing many services through Bluemix that can be easily integrated into applications through a REST API. There are many services available on Watson, and they can be divided into major categories such as “language”, “speech”, “vision”, and “data insights”. Some very popular services include Natural Language Understanding for recognizing the meaning behind natural language and Visual Recognition for classifying contents in images. This adventure focuses on Natural Language Understanding using Watson’s Conversation service.
> 
> Because interfacing with a REST endpoint using low-level networking calls quickly becomes inefficient, there are many Watson Developer Cloud SDK’s that have been created to simplify making requests, exposing natural functions and data structures for the language. What would have taken over 500 lines of code, takes now 10. The Watson Swift SDK, previously named the iOS SDK, is now the second most used Watson SDK (after the NodeJS SDK). It had its origins, obviously, for the iOS mobile use case, but since the Swift programming language now can be used for running Linux, it has opened the SDK for usage for server-side programming as well. Now you can use the same SDK you have bundled in your client application, and easily move that functionality out to the server, giving you more control over the frequency at which calls to Watson are made, have a single point for credentials for making the calls, and also can create applications where the result of the request to Watson can be shared with multiple users.
> ```

since the Swift programming language can now be used for running Linux

used running _on_ Linux ? 
or maybe just 
used on Linux?

Also.. heads up that you'll want to soft-line break the file at some arbitrary column number, not because it's needed for display, but because it makes merges and changes easier to deal with in github. Eg, comments can only be attached per 'line' in the file, and this line is an entire paragraph, making it harder to handle for merges etc in future.

---

*[walkthroughs/addWatsonConversation.adoc, line 30 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV1MyQHa02xyKqJpdt:-KgV1MyQHa02xyKqJpdu:bpx86cf) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L30)):*
> ```AsciiDoc
> First, you’ll need to create and configure a new Watson Conversation service. Building out a new Conversation service should only take a few minutes once you have a Bluemix account (you can get a free test account too).
> 
> . Create a new Watson Conversation service in link:https://console.ng.bluemix.net[Bluemix].
> ```

take a peek at the other adventure adoc source to see how to move the urls to the top of the doc making it easier to migrate them in future if we have to

---

*[walkthroughs/addWatsonConversation.adoc, line 94 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4XABMktlRRQqtFgq:-KgV4XABMktlRRQqtFgr:b6iict6) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L94)):*
> ```AsciiDoc
> ## Walkthrough
> 
> . Fork and clone the sample-swift-room
> ```

throw in some short words here about how to do the fork.. you've got the clone covered via the command below. Can't assume people following this have ever used git/github before, so we have to be kind.

---

*[walkthroughs/addWatsonConversation.adoc, line 104 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4hINuJN62nISgjec:-KgV4hINuJN62nISgjed:b-hbmbq3) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L104)):*
> ```AsciiDoc
> . Add the Watson SDK as a dependency in your project
> +
> In your Package.swift file, add to the bottom of the list of dependencies the following library:
> ```

consider linking the package.swift in the sample repo here, so people can see where it is within the sample.. and have an easy way to view the original one.

---

*[walkthroughs/addWatsonConversation.adoc, line 112 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4rl-wzp8IJehp5EJ:-KgV4rl-wzp8IJehp5EK:b-4t2hc) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L112)):*
> ```AsciiDoc
> 
> [start=3]
> . In Source/SwiftRoom/RoomImplementation.swift, import the Conversation service modules:
> ```

maybe explain a little about where these come from, if they have api docs, link them here, give people the power to go learn more.

---

*[walkthroughs/addWatsonConversation.adoc, line 127 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV1xIJSdcYH_o5kMTy:-KgV1xIJSdcYH_o5kMTz:bwjn7ty) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L127)):*
> ```AsciiDoc
> let password = "password-goes-here"
> let version = "2017-03-22”
> let workspaceID = "workspace-id-goes-here"
> ```

Assuming this is the workspace-id from prereqs.. it needs some more text in this section to explain where this, and the username/password values come from (again likely also from prereqs, but maybe they only skimmed that bit, and hit here).. and explain why version is set at this value.. The user is supposed to be learning a little about why they are doing stuff, rather than being a cut & paste robot

---

*[walkthroughs/addWatsonConversation.adoc, line 131 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5BnDeKtQsSoNINcq:-KgV5BnDeKtQsSoNINcr:b7hkcj2) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L131)):*
> ```AsciiDoc
> 
> [start=5]
> . Add a Conversation service object to your RoomImplementation class:
> ```

add short description of what a Conversation service object is.. quote it from the documentation if you wish.. someone reading this needs to get the 'why' behind why they are adding this object, what does it enable them to do etc.. 

---

*[walkthroughs/addWatsonConversation.adoc, line 143 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5SDBJJdN71su0Nil:-KgV5SDBJJdN71su0Nim:b-5kebm8) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L143)):*
> ```AsciiDoc
> . Add a Context for the conversation
> +
> A context allows the conversation to hold state. For instance, when you are finished selecting the meat on your sandwich, Watson continues the conversation with asking about toppings. Since REST calls are stateless by nature, we can hold state by keeping a state ID that is updated whenever you get a new response back.
> ```

This bit is good.. s'kinda what's missing for the Conversation object 

---

*[walkthroughs/addWatsonConversation.adoc, line 151 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV2KyQ16e8-LIEtDmT:-KgV2KyQ16e8-LIEtDmU:b9uj7fh) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L151)):*
> ```AsciiDoc
> 
> [start=7]
> . Set the context of the conversation on application start.
> ```

why? needs more words.. ?
Maybe it's just because I like don't know any swift, but how did the declaration of var context tie into self.context = response.context ? I'm guessing "response in" is somehow related to the response.context? I don't know though, because it's left to me to figure out ;)

---

*[walkthroughs/addWatsonConversation.adoc, line 168 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5cPXyeemr2KEtnYY:-KgV5cPXyeemr2KEtnYZ:b5rl0ud) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L168)):*
> ```AsciiDoc
> 
> [start=8]
> . RoomImplementation.swift will now look like this:
> ```

if you have a branch of your room where the changes are already made, here's a great place to link that branch, so people can go browse the code (or even clone the code if their lazy/busy etc) directly.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-KgV6tM74vN5BN9Hpocd:b-8iikb8) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
> ```AsciiDoc
>                 }
> 
>                 try endpoint.sendMessage(connection: connection,
> ```

you may want to do the user's chat message send before you do the watson reply send.. otherwise I suspect chats will be backwards.. eg

Watson: Sunny.
You: Whats the Weather Like?

---

*[walkthroughs/addWatsonConversation.adoc, line 243 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5rfV7zLI3JWq3Sv9:-KgV5rfWznp1B6oaOurj:b-ck72ot) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L243)):*
> ```AsciiDoc
> [start=10]
> . Compile your project with
> +
> ```

be careful here, the last command they ran was git clone which leaves them in the directory above the extracted files.. you may want to throw something in to tell them to change directory to the same dir as file blah before they do this bit.

---

*[walkthroughs/addWatsonConversation.adoc, line 255 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV63FjfgW8sRM9Zvio:-KgV63FkVceRHrIOj3_u:b-nlsmlk) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L255)):*
> ```AsciiDoc
> [source]
> ----
> $ .build/debug/GameOn
> ```

does swift run on windows? using git-bash? (this path is unix style, so I'm guessing git-bash will likely be able to handle it)

---

*[walkthroughs/addWatsonConversation.adoc, line 259 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3COTAS1BlbzCsDdA:-KgV3COTAS1BlbzCsDdB:b-2zuqm7) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L259)):*
> ```AsciiDoc
> 
> Then access http://localhost:8080/ in your browser. Visiting this page provides a small chat window you can use to test Watson in your service directly.
> ```

Do you explain anywhere how to host the room in bluemix, or build it as a container (I'm guessing back in your create a swift room doc), so here you can tell people how they can test the room by running it in bluemix, and using `/sos` `/teleport myRoomName` to goto their room, and interact directly with it via the live game?

---

*[walkthroughs/addWatsonConversation.adoc, line 263 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3Uk933wFha12KEwZ:-KgV3UkAeMvouHjERyEB:b-pcqah5) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L263)):*
> ```AsciiDoc
> ## Suggested extensions
> 
> The more the Conversation service is developed, the more interesting this adventure will be. Use your imagination!
> ```

Here give some examples, regardless of how crazy.. =) 

Possibly talk about the other watson api's and how they could be explored as extensions to using the conversation api

---

*[walkthroughs/addWatsonConversation.adoc, line 267 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3iMESrkS7ySCqdmg:-KgV3iMESrkS7ySCqdmh:bsvkd3r) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L267)):*
> ```AsciiDoc
> ## Conclusion
> 
> We hope you enjoyed this adventure and feel inspired to explore your GameOn Swift room with Watson.
> ```

Mention here the things they learned during the walkthrough.. the basics of setting up a service, and invoking it from swift. The whole 'tell them what you will tell them, tell them what you tell them, tell them what you told them' pattern.. this is the last part where you summarise the basic skills they acquired via going through the tutorial part. 

There should also be a further adventures section, to link to other adventures they could work through.. admittedly this is currently the only swift adventure, so it's hard to say too much.. 

---

*[walkthroughs/createMore.adoc, line 87 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgUzopB1fLXCxeoSDsK:-KgUzopB1fLXCxeoSDsL:b-9mbhbw) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/createMore.adoc#L87)):*
> ```AsciiDoc
> === Watson
> 
> Want to add a Watson bot to your chatroom? Integrate Watson’s Conversation service into your GameOn Swift room.
> ```

This part feels badly placed (in the middle of the fault tolerance block) .. it would likely be better placed further up after the persistence block, and it should use a languagename swift tag to mark it as swift content with a link through to the tutorial.. 

It should probably be renamed from Watson to "Cognitive and Analytics" =) the link to the adventure doc can mention watson, if it needs to, but should try to stay short & 1 line ish to blend in with the other stuff. 

Keep the paragraph for the section brief, it'll need to cover any future usage of watson, or other analytics and stuff from any language =)

I'm guessing this para came from a merge to an earlier version of this page before it got restructured to make it more index like =)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

rfdickerson commented:
Thanks for the good comments here. Just FYI @jkingoliver is out on vacation in Spain right now. She should be able to get your edits in next week.

jkingoliver commented:
Review status: 8 of 11 files reviewed at latest revision, 17 unresolved discussions.

---

*[walkthroughs/addWatsonConversation.adoc, line 10 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV03U5l7mjmXtPBVfv:-KhYq68h4VOeQ6Ewi7LZ:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L10)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

learn about how to integrate the Watson Conversation service
or
learn how to integrate Watson Conversation service

Something doesn't sit right with it worded as is.. I think it's the 'about' that seems superfluous .. 
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 22 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV0cL0Y1rnX5Vv9V6o:-KhYsCI2cx_y39ci_4lt:b-xe8r64) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L22)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

since the Swift programming language can now be used for running Linux

used running _on_ Linux ? 
or maybe just 
used on Linux?

Also.. heads up that you'll want to soft-line break the file at some arbitrary column number, not because it's needed for display, but because it makes merges and changes easier to deal with in github. Eg, comments can only be attached per 'line' in the file, and this line is an entire paragraph, making it harder to handle for merges etc in future.
</blockquote></details>

I'm going with this:

"since the Swift programming language now can be used on Linux"

---

*[walkthroughs/addWatsonConversation.adoc, line 30 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV1MyQHa02xyKqJpdt:-KhZ-ZGl6t6VrZvh24bI:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L30)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

take a peek at the other adventure adoc source to see how to move the urls to the top of the doc making it easier to migrate them in future if we have to
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 94 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4XABMktlRRQqtFgq:-KhZ0COS2cJHeO_2Z_Uc:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L94)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

throw in some short words here about how to do the fork.. you've got the clone covered via the command below. Can't assume people following this have ever used git/github before, so we have to be kind.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 104 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4hINuJN62nISgjec:-KhZ0qKriqbH2ic9mDGC:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L104)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

consider linking the package.swift in the sample repo here, so people can see where it is within the sample.. and have an easy way to view the original one.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 112 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4rl-wzp8IJehp5EJ:-Khbz6ZToElYUXY85tjN:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L112)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

maybe explain a little about where these come from, if they have api docs, link them here, give people the power to go learn more.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 127 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV1xIJSdcYH_o5kMTy:-Khc0BO8_VKhT8L7dDhu:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L127)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

Assuming this is the workspace-id from prereqs.. it needs some more text in this section to explain where this, and the username/password values come from (again likely also from prereqs, but maybe they only skimmed that bit, and hit here).. and explain why version is set at this value.. The user is supposed to be learning a little about why they are doing stuff, rather than being a cut & paste robot
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 131 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5BnDeKtQsSoNINcq:-Khc224V0G629Ua89OR6:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L131)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

add short description of what a Conversation service object is.. quote it from the documentation if you wish.. someone reading this needs to get the 'why' behind why they are adding this object, what does it enable them to do etc.. 
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 151 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV2KyQ16e8-LIEtDmT:-Khc4xp3uXKyxyeubyGl:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L151)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

why? needs more words.. ?
Maybe it's just because I like don't know any swift, but how did the declaration of var context tie into self.context = response.context ? I'm guessing "response in" is somehow related to the response.context? I don't know though, because it's left to me to figure out ;)
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 168 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5cPXyeemr2KEtnYY:-Khc5RN1HGazd4tSOwtz:bnykq6c) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L168)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

if you have a branch of your room where the changes are already made, here's a great place to link that branch, so people can go browse the code (or even clone the code if their lazy/busy etc) directly.
</blockquote></details>

I'll try to create one, but don't have a branch to link atm.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-Khc5cP-_GIB23NDiSfS:bqityfa) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

you may want to do the user's chat message send before you do the watson reply send.. otherwise I suspect chats will be backwards.. eg

Watson: Sunny.
You: Whats the Weather Like?
</blockquote></details>

I don't remember seeing out of order chats, but this is a good point...I'll re-investigate.

---

*[walkthroughs/addWatsonConversation.adoc, line 243 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV5rfV7zLI3JWq3Sv9:-Khc6htVeSNOYFuB8yeA:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L243)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

be careful here, the last command they ran was git clone which leaves them in the directory above the extracted files.. you may want to throw something in to tell them to change directory to the same dir as file blah before they do this bit.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 255 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV63FjfgW8sRM9Zvio:-Khc6slt42Va-xMlbj9F:b-s6yo2z) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L255)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

does swift run on windows? using git-bash? (this path is unix style, so I'm guessing git-bash will likely be able to handle it)
</blockquote></details>

Nope, only macos and Linux.

---

*[walkthroughs/addWatsonConversation.adoc, line 259 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3COTAS1BlbzCsDdA:-KhcLogMPsnyVAefkeDe:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L259)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

Do you explain anywhere how to host the room in bluemix, or build it as a container (I'm guessing back in your create a swift room doc), so here you can tell people how they can test the room by running it in bluemix, and using `/sos` `/teleport myRoomName` to goto their room, and interact directly with it via the live game?
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 263 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3Uk933wFha12KEwZ:-KhcSOvcYpSn4W5dkAlh:b-p70dle) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L263)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

Here give some examples, regardless of how crazy.. =) 

Possibly talk about the other watson api's and how they could be explored as extensions to using the conversation api
</blockquote></details>

Adding suggestion to connect additional services such as Tone Analyzer and Speech to Text.

---

*[walkthroughs/addWatsonConversation.adoc, line 267 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3iMESrkS7ySCqdmg:-KhcZNVXd1IINjkC8DB7:b-cwfqsh) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L267)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

Mention here the things they learned during the walkthrough.. the basics of setting up a service, and invoking it from swift. The whole 'tell them what you will tell them, tell them what you tell them, tell them what you told them' pattern.. this is the last part where you summarise the basic skills they acquired via going through the tutorial part. 

There should also be a further adventures section, to link to other adventures they could work through.. admittedly this is currently the only swift adventure, so it's hard to say too much.. 
</blockquote></details>

Ok, for a "Suggested further adventure", I suggest persistence via cloudant and point to the Java advanced adventure. I also point to one of the Swift team's sample applications, TodoList-CouchDB, to show how to connect to a cloudant instance using the Swift language.  Kind of a stretch, but it's something...

---

*[walkthroughs/createMore.adoc, line 87 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgUzopB1fLXCxeoSDsK:-KhcactJe0yHBW6WDS3V:bxfg90n) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/createMore.adoc#L87)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

This part feels badly placed (in the middle of the fault tolerance block) .. it would likely be better placed further up after the persistence block, and it should use a languagename swift tag to mark it as swift content with a link through to the tutorial.. 

It should probably be renamed from Watson to "Cognitive and Analytics" =) the link to the adventure doc can mention watson, if it needs to, but should try to stay short & 1 line ish to blend in with the other stuff. 

Keep the paragraph for the section brief, it'll need to cover any future usage of watson, or other analytics and stuff from any language =)

I'm guessing this para came from a merge to an earlier version of this page before it got restructured to make it more index like =)


</blockquote></details>

Oops! 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

rfdickerson commented:
Excellent! Thanks @jkingoliver 👍 Does this resolve everything needed to get this merged in except for a link to a working implementation?

jkingoliver commented:
@BarDweller Let me know if we can merge this after another look-over. Thanks!

jkingoliver commented:
I added a link to a sample my personal github account.

ebullient commented:
Reviewed 1 of 11 files at r2, 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

BarDweller commented:
Reviewed 1 of 11 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-Kia6PpRvfufs8iB9gw8:bjxb9qe) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

I don't remember seeing out of order chats, but this is a good point...I'll re-investigate.
</blockquote></details>

is it only sending the Watson message now, and not the user's chat message? 

---

*[walkthroughs/addWatsonConversation.adoc, line 259 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3COTAS1BlbzCsDdA:-Kia6wkiaat9DYl0aS87:bsp9x0z) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L259)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm not sure there's enough here still.. the existing deploy docs work for java and assume you are either a CF App, or have built a docker image that will run your room. It's not clear to me from the instructions you have that you end up with either of those, in which case a user is still left to figure that out?

---

*[walkthroughs/addWatsonConversation.adoc, line 267 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3iMESrkS7ySCqdmg:-Kia7I_yne88ZB6uY0PG:b-k9eja0) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L267)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Ok, for a "Suggested further adventure", I suggest persistence via cloudant and point to the Java advanced adventure. I also point to one of the Swift team's sample applications, TodoList-CouchDB, to show how to connect to a cloudant instance using the Swift language.  Kind of a stretch, but it's something...
</blockquote></details>

Sure, if/when we have additional swift adventures, we can always add links to them then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

jkingoliver commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[walkthroughs/addWatsonConversation.adoc, line 10 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV03U5l7mjmXtPBVfv:-KhcfmPssG2rtLnox8VE:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L10)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 22 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV0cL0Y1rnX5Vv9V6o:-Khcfo4UBEyOQ2G6OrrZ:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L22)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

I'm going with this:

"since the Swift programming language now can be used on Linux"
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 30 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV1MyQHa02xyKqJpdt:-Khcfp3BtomcSDNJfP8-:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L30)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 94 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV4XABMktlRRQqtFgq:-KhcfqAi3U8aO-BcgB4Y:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L94)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-Kia83iWelC9y9eNNUy0:b-8xtx8t) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

is it only sending the Watson message now, and not the user's chat message? 
</blockquote></details>

Yes, that's right. There was a typo in the docs that I corrected in the sample.

---

*[walkthroughs/addWatsonConversation.adoc, line 259 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3COTAS1BlbzCsDdA:-Kia9qd9obltzRyIxbhP:b-g86wg) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L259)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

I'm not sure there's enough here still.. the existing deploy docs work for java and assume you are either a CF App, or have built a docker image that will run your room. It's not clear to me from the instructions you have that you end up with either of those, in which case a user is still left to figure that out?
</blockquote></details>

Section 12 is what I added and it points to the docs that describe how to deploy using cf, which is how the swift room app would be deployed to bluemix. I'm not sure how to make it more clear. What question exactly do you need answered?

---

*[walkthroughs/createMore.adoc, line 87 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgUzopB1fLXCxeoSDsK:-Ki1M7icB6r5o21jIzgK:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/createMore.adoc#L87)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Oops! 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

BarDweller commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-KiaAVK_kkAGN4FbSTxP:b-atjtyg) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Yes, that's right. There was a typo in the docs that I corrected in the sample.
</blockquote></details>

For Game On, when the user sends you a chat message, your room has to send the response that makes the chat message visible to all the other room participants.. with this as-is, everyone else will see the "Watson: response" but not the "You: question" part .. Which will get a bit confusing.

---

*[walkthroughs/addWatsonConversation.adoc, line 259 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3COTAS1BlbzCsDdA:-KiaBPwvbwGS-JNNk9dO:b-gjr7d4) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L259)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Section 12 is what I added and it points to the docs that describe how to deploy using cf, which is how the swift room app would be deployed to bluemix. I'm not sure how to make it more clear. What question exactly do you need answered?
</blockquote></details>

Thanks for the clarification re cf, aye, lets link to the cf deploy doc rather than the generic one, unless you fancy writing up how to make it work with the container service too ;p

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

jkingoliver commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[walkthroughs/addWatsonConversation.adoc, line 259 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV3COTAS1BlbzCsDdA:-KiaDS1-GzuDfiUAaYZ1:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L259)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

Thanks for the clarification re cf, aye, lets link to the cf deploy doc rather than the generic one, unless you fancy writing up how to make it work with the container service too ;p
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

jkingoliver commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-KiaDWPHMec_z321c5Co:b-9xf4zg) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
<details><summary><i>Previously, BarDweller (Ozzy Osborne) wrote…</i></summary><blockquote>

For Game On, when the user sends you a chat message, your room has to send the response that makes the chat message visible to all the other room participants.. with this as-is, everyone else will see the "Watson: response" but not the "You: question" part .. Which will get a bit confusing.
</blockquote></details>

Oh right... I will fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

jkingoliver commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[walkthroughs/addWatsonConversation.adoc, line 233 at r3](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65#-KgV6tM74vN5BN9Hpocc:-KiaJHFaxJig5B1XbXD7:b-896fix) ([raw file](https://github.com/gameontext/gameon-gitbook/blob/8d420b1ae5d9ec2807ac348799d2fab283f7371d/walkthroughs/addWatsonConversation.adoc#L233)):*
<details><summary><i>Previously, jkingoliver (Jennifer Oliver) wrote…</i></summary><blockquote>

Oh right... I will fix.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->

BarDweller commented:
Reviewed 1 of 2 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/gameontext/gameon-gitbook/65)*
<!-- Sent from Reviewable.io -->
